lib/puppet/indirector/catalog/compiler.rb

```diff
@@ -53,7 +53,7 @@ def find(request)
     node.trusted_data = Puppet.lookup(:trusted_information) { Puppet::Context::TrustedInformation.local(node) }.to_h
 
     if node.environment
-      node.environment.with_text_domain do
+      node.environment.with_guarded_environment do
```

Comment:
We can't store guarded state on the Environment itself, because it may be discarded and reloaded due to the ttl. The state needs to live with the Entry stored in the cache.